### PR TITLE
Take first stab on project lifecycle

### PR DESCRIPTION
**Pull request by fdegir**
### Applicable Issues

https://github.com/eiffel-community/community/issues/5
https://github.com/eiffel-community/eiffel-repository-template/issues/9

### Description of the Change

Eiffel Community is organized around Git repositories and how those repositories relate to each other is not very clear for user perspective since one repository may be part of a project which consists of multiple repositories. In order for the community to determine what projects are hosted, the community first needs to document the fact that EIffel Community is organized around projects which may well consist of multiple repositories.

In addition to how the community is organized, there is no easy way for the community and users to know what stage those projects are in their lifecycle. This is an important aspect to know for users to decide whether they should use the project in production or not.

This PR proposes a simple project lifecycle which includes
- what Eiffel Community is organized around
- what is the lifecycle of the projects hosted under Eiffel Community
- how a new project can be created and requirements to do so
- what are the different stages Eiffel Projects are placed in
- how projects could move between different stages

### Benefits

Having information about the status of projects upfront makes it easier for potential users to estimate the effort needed for deployment.

### Sign-off

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

Signed-off-by: Fatih Degirmenci fatih.degirmenci@est.tech

### CODE REVIEW

**Comment by e-backmark-ericsson** [comment]
Looks promising :)
I'm a bit disturbed by the stage name 'Graduated'. Wouldn't it be simpler to call it 'Mature' or 'Active'?
Also, how/where would we document a projects status? There is no 'project' entity in Github matching this use case as far as I'm aware. Should we groupd repos in a project using Github labels, or should we request each project to document themselves on https://eiffel-community.github.io/ or somewhere else?

**Comment by fdegir** [comment]
> I'm a bit disturbed by the stage name 'Graduated'. Wouldn't it be simpler to call it 'Mature' or 'Active'?

I don't remember where but someone said "mature" may give wrong impression with regards to the projects that are not marked as mature; are the others immature? "Active" is similar to this - does it imply that the others are not active? Graduated is kind of a middle ground and fits well if we take it together with the requirement to be moved to graduated - at least one production use. So people can see that "it is used in production so that's why it is graduated" or something like this.

Btw, I'm fine with all options as long as we make what we mean with those stages clear and understandable so the people can take conscious decisions when it comes to how comfortable they could feel when they intend to use a certain project in production.

> Also, how/where would we document a projects status? 

I think we could have some kind of badge on the repos as discussed on [this issue](https://github.com/eiffel-community/eiffel-repository-template/issues/9). In addition to that, the community maintainers can maintain project list in this repo where we can list which repos are considered as part of which project, maturity level of the projects and so on. We can even add a step to project creation/review process asking maintainers to update the table with a PR as part of the proposal.

> There is no 'project' entity in Github matching this use case as far as I'm aware. Should we groupd repos in a project using Github labels, or should we request each project to document themselves on https://eiffel-community.github.io/ or somewhere else?

I think that is not very important. The idea with "projects" is to make things a bit more visible - both about how things are structured and how mature things are. So again, having a document to list projects and the repos owned by those projects would be a good start. More can be investigated later on if necessary.

**Comment by e-backmark-ericsson** [comment]
Argumentation about naming accepted. I could be fine with several names on the stages as long as it is clear what they mean.

About having a document in this community repo listing all projects and their repositories makes me concerned about loosing synch between that document and the actual repositories. I believe that that document should just list the top/main repository for each project and then each top/main repository should list all its constituent repos instead. Badges representing the project status could be good. Putting them on each repository is probably a good idea, even though there could be a synch issue there as well as we need to remember to update the community document with status as well as the badges in all project repos. But let's give it a try.

Later on we should probably consider how this project/repo listing relates to [Sepia](https://eiffel-community.github.io/eiffel-sepia/) as well.

File: PROJECT_LIFECYCLE.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+and archived. In addition to the project lifecyle process, stages which the projects can be admitted or transitioned
+into as well as criteria and expectations for the stages are described.
+
+## Projects
+
+The activities of the Eiffel Community are organized into projects targeting different areas within the scope of
+Eiffel Community including but not limited to development and maintenance of Eiffel Protocol, development of new
+functionality that adheres to the protocol, integration of other technologies, testing, documentation, as well as
+the support functions to create and maintain the infrastructure and the toolchain community uses.
+
```
Comment (e-backmark-ericsson) [inline PROJECT_LIFECYCLE.md:26]: Refer to PROJECTS.md here, where the current list of projects can be seen?

File: PROJECT_LIFECYCLE.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+Community repository by creating an issue.
+
+### Project Acceptance Process
+
+Project review and acceptance happen directly on the issue asking for project creation. If community members
+require further discussions, ad-hoc meetings can be scheduled to discuss the proposal further but this is not
+mandatory.
+
+Simple majority of Eiffel Community Maintainers is required for a project to be accepted to Eiffel Community.
+
```
Comment (e-backmark-ericsson) [inline PROJECT_LIFECYCLE.md:50]: Add that a PR updating PROJECTS.md will be done within the acceptance process as well (?).

Comment (fdegir) [reply PROJECT_LIFECYCLE.md:52]: done

File: PROJECT_LIFECYCLE.md (hunk 1 above)
Comment (fdegir) [reply PROJECT_LIFECYCLE.md:26]: done

**Comment by t-persson** [comment]
An addition to the stage name discussion. I kinda like the "Development status" classifiers for python projects in pypi.
Does not mean we have to use all of them, but it's worth looking at. I particularly like that they separated Production/Stable and Mature where "Mature" can be seen as "Yea we won't make any drastic changes anymore, but it's used in production". 

Development Status :: 1 - Planning
Development Status :: 2 - Pre-Alpha
Development Status :: 3 - Alpha
Development Status :: 4 - Beta
Development Status :: 5 - Production/Stable
Development Status :: 6 - Mature
Development Status :: 7 - Inactive

I'm fine with the current stages in this PR

**Comment by fdegir** [comment]
> About having a document in this community repo listing all projects and their repositories makes me concerned about loosing synch between that document and the actual repositories.

That is always a possibility so making it part of the project creation/review process could reduce the chances of it to happen. We will see how well this process goes.

> I believe that that document should just list the top/main repository for each project and then each top/main repository should list all its constituent repos instead. 

I think this makes a lot of sense so the responsibility to document project status (stage, repos, maintainers, etc.) fall on the project maintainers and reduce the need for syncing changes back and forth between project repos and the community repo.

> Badges representing the project status could be good. Putting them on each repository is probably a good idea, even though there could be a synch issue there as well as we need to remember to update the community document with status as well as the badges in all project repos. But let's give it a try.

Making it part of project review process will hopefully help here.

**Comment by fdegir** [comment]
> the "Development status" classifiers

That makes sense but it can perhaps be left to the projects to have even further status information about the state of their project. The reason for this is that the development status requires more knowledge about the status of things and it may become difficult for community maintainers to judge that during project review.

The stages documented in the community repository and applied to the projects mean that the community "blessed"  the stage the project is in based on the input presented by the projects such as production use, health, etc which can easily be seen by community maintainers.

File: PROJECT_LIFECYCLE.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+Project maintainers are expected to update PROJECTS.md file by sending a pull request to community repository
```
Comment (e-backmark-ericsson) [inline PROJECT_LIFECYCLE.md:53]: Please make a link to PROJECTS.md

Comment (fdegir) [reply PROJECT_LIFECYCLE.md]: done

**Comment by sodero** [comment]
Looks good to me. Excellent work.

**Comment by fdegir** [comment]
I suppose we can wait one more day and merge this tomorrow so we can start populating list of projects in PROJECTS.md.

**Comment by fdegir** [comment]
submitting this. thanks for reviews.

**Pull request merged**